### DISCUSSION TITLE
Why two microcontrollers?

### DISCUSSION THREAD

**Discussion by pilotniq** [Arnold-n/P1P2MQTT#60] [Q&A] [answered]
I'm learning about this project, and I'd like to understand why the hardware device needs both an ATMega 328 and an ESP8122.

Do you think an ESP32 could do both the P1P2 serial decoding (that the ATMega currently does) and the WiFi interfacing? I assume that it would be simpler to maintain one code base, and also for firmware updates.

Background: I'm considering getting a Daikin Altherma Geo 3 heatpump. I'd like to implement my own wireless room temperature sensor solution instead of Daikin's wired Madoka or ugly EKRTR wireless room sensor.

**Comment by Arnold-n** [comment, marked as answer]
The reason for a 2-CPU solution is both history and preference. History as the bus decoding was developed on an Arduino Uno, and an ESP was added later to do the high-level decoding. And preference as a 2-CPU solution is architecturally preferred to have low-level hard-real-time code on a hard real-time ATmega, and to have high-level code on a soft-realtime ESP. As the hardwareserial library on the ESP also indicates that it is inherent not error-free,  it may be difficult to achieve the same timing reliability on an ESP as on the ATmega.

The current solution does not replace the Madoka, but supplements it, just like a (W)LAN adapter.

**Reply by pilotniq** [reply]
Thank you for the explanation!

Couldn't the existing solution emulate the Madoka's way of for instance reporting the room temperature to the heat pump? Thereby making the Madoka unnecessary if the only desired functionality as a room temperature sensor.

**Reply by Arnold-n** [reply]
Depends - I assume the Madoka operates as main/master controller and we do not (yet) plan to overtake that action. If the Madoko operates as sub/auxiliary/slave controller to a master (perhaps in the indoor unit) then it can easily be replaced.

Also, there is a menu option for an "external room temperature sensor" but I am not sure yet how that works.

**Reply by javimurcia** [reply]
> Also, there is a menu option for an "external room temperature sensor" but I am not sure yet how that works.

That's something I can answer, at least for my particular unit (erla, I'll update the answer when I get home with the full model number)

The interior unit has 2 terminals labeled  external temperature sensor. It's just a $150 thermistor, and luckily, one of the service manuals have the temperature-resistance chart. 22k is read as 23°C

After installing it, you have the option of setting said temperature sensor as an exterior temperatures probe, or an internal temperatura probe.

If you choose exterior, it's the reading of that probe (instead of the temperatura reported by the exterior unit) that will be used for the climatic compensation curve.

If you choose interior, then that probe is used for measuring the ambient temperature.

Problem, for that to be of any use, you have to choose "thermostat control" as the main zone mode of control, which require having a madoka thermostat installed. And if you enable it without a madoka wired, you will get an error (communication error with the room thermostat) a few seconds after startup.

> Depends - I assume the Madoka operates as main/master controller and we do not (yet) plan to overtake that action. If the Madoko operates as sub/auxiliary/slave controller to a master (perhaps in the indoor unit) then it can easily be replaced.

From what I have read on your documentation, the p1p2 bus can only have a single master. In that case, I highly doubt the madoka will act as the bus master, as I assume that role is most likely be fulfilled by the indoor unit's interface (which if I'm not mistaken, communicates over p1p2 with the hydro pcb) and also because the madoka is an optional component.

Anyway, that's my end game, emulate a madoka so I can report whatever room temperature I want to the heat pump and modulate the leaving water temperature.

Right now I'm trying to find a cheap madoka so I can start to thinker 

**EDIT:** you can find the temperature - resistance table for the external thermistor at page 262 of the Daikin Altherma 3 M service manual, available at https://www.manualslib.com/manual/2560305/Daikin-Altherma-3-M.html?page=262

**Reply by Arnold-n** [reply]
Many thanks for your explanation. I wonder if all indoor units have their interface connected via P1/P2 - some may use a separate connection. But if Madoka can communicate the room temperature as aux controller, we should be able to do the same. On Dutch  www.marktplaats.nl I regularly see new Madoka offered for around Euro 50 from installations where they are simply not used - I have one here but no time yet to play with it.

**Reply by javimurcia** [reply]
Which version do you have? Because I have seen cheap madokas that are BRC1H52x, that are for ac units.

In my unit's documentation, the only compatible madoka is the BRC1HHDx

And both are physically identical, but I'm pretty sure that the cheap one is not going to work. And it sucks spending 180€ for something you intend to replace.

I'll keep looking.

BTW, my interior's unit user interface appears to comunícate over P1P2
https://drive.google.com/file/d/1L0TwRcmHT9wNZw5N-vzALTd8-iHajakm/view?usp=drivesdk

https://drive.google.com/file/d/19iysUoingXjmUnvPjaGN1WAmUKXDC1at/view?usp=drivesdk

**Reply by Arnold-n** [reply]
Mine is a BRC1HHDAW, perhaps I was just lucky. In your situation I agree that the Madoka will likely be an aux controller. Even then I am not yet sure whether the Madoka acts exactly in the same way as other (LAN type) aux controllers, as newer systems seem to have different addresses for different controllers (30, 31, and 3F), whereas older systems only supported 30 (or sometimes 30 and 31). If you want to monitor the Madoka traffic with my interface to find how to emulate it  I can lend you my Madoka.

**Reply by javimurcia** [reply]
It has to have some mechanism to address multiple devices, since at least, you can connect a madoka and a cascade controller adapter to the p1/p2 bus.

The traditional wireless wired gateway appears to be incompatible (at least it is not in the accessory list), and the cloud functionality is archived via an SD-like wifi adapter, or a wireless gateway that appears to communicate over uart.

I'm still building the p1p2 interface, but I'll let you know when I have it up and running, if I still haven't sourced a reasonably priced madoka, so I can sniff its traffic and try to emulate it.

**Reply by javimurcia** [reply]
> Mine is a BRC1HHDAW

The plot thickens: according to the compatibility table of my unit, the reference for the madoka remote controll should be BRC1HHDx (note the missing A, the x is the color: W white, K black, and S silver).

And of course, daikin offers no explanation on what are the differences...

**Reply by gmcmicken** [reply]
My guess is "A" might be "North America" since the hits on search for that model are in English, and the other model results are all different languages.
